### DISCUSSION TITLE
Understanding how Radix Primitives merge props onto passed in children

### DISCUSSION THREAD

**Discussion by wencyen** [radix-ui/primitives#2382] [Help] [answered]
Hi all, I've been wondering about how `radix ui` merges the primitive props of `Trigger` with that of `IconButton`'s `<button>` element. This is not a question of how to use the interface, but rather where in the source code does this magic exist!? I looked at the implementation of `DropdownMenuTrigger` [here](https://github.com/radix-ui/primitives/blob/main/packages/react/dropdown-menu/src/DropdownMenu.tsx#L95)  but was unable to find anything about how it is able to merge props onto the children in this case `IconButton`. Thanks!
```js
import { MoreHoriz as MoreHorizIcon } from '@mui/icons-material'
import { IconButton } from '@mui/material'
import * as DropdownMenu from '@radix-ui/react-dropdown-menu'
import { PropsWithChildren, ReactNode } from 'react'
const MenuDemoWithRadix = () => {
  <DropdownMenu.Root>
    <DropdownMenu.Trigger asChild>
      <IconButton aria-label="Toggle menu" size="medium">
        <MoreHorizIcon fontSize="medium" />
      </IconButton>
    </DropdownMenu.Trigger>
    // ... rest of the component
  </DropdownMenu.Root>
}
```
![Screen Shot 2023-09-06 at 2 54 32 PM](https://github.com/radix-ui/primitives/assets/49729798/ed20159c-730c-48de-b317-0d43a9ac7d27)

**Comment by joaom00** [comment, marked as answer]
Hey @codepath2019, the magic happens in the [`Slot`](https://github.com/radix-ui/primitives/blob/main/packages/react/slot/src/Slot.tsx) component and Radix uses it through [`Primitive`](https://github.com/radix-ui/primitives/blob/main/packages/react/primitive/src/Primitive.tsx#L47) helper

**Reply by wencyen** [reply]
thanks for responding @joaom00 I will look into the source of this magic
